contracts/test/StakeTableInvariant.t.sol

```diff
@@ -0,0 +1,341 @@
+
+        // Exit if it is too early to deposit
+        if (stakeTable.currentEpoch() <= node.registerEpoch) {
+            return;
+        }
+
+        // Some exit is in progress, do not deposit
+        if (node.exitEpoch != 0) {
+            return;
+        }
+
+        amount = uint64(bound(amount, 1, 10));
+
+        // Exit if the tokenCreator ran out of funds
+        if (token.balanceOf(tokenCreator) < amount) {
+            return;
+        }
+
+        // Transfer some tokens to userAddress
+        vm.prank(tokenCreator);
+        token.transfer(node.account, amount);
+
+        // Prepare for the token transfer
+        vm.prank(node.account);
+        token.approve(address(stakeTable), amount);
+
+        vm.prank(node.account);
+        stakeTable.deposit(vk, amount);
+    }
+
+    function withdrawFunds(uint256 rand) public {
+        // Check if some withdrawals are possible
+        if (requestExitKeys.length == 0) {
+            return;
+        }
+
+        uint256 index = bound(rand, 0, requestExitKeys.length - 1);
+
+        BN254.G2Point memory vk = requestExitKeys[index];
+        bytes32 vkHash = stakeTable._hashBlsKey(vk);
+        uint64 exitEpoch = exitEpochForBlsVK[vkHash];
+
+        // Ensure we have reached the right epoch before withdrawing
+        uint64 slackForEscrowPeriod = 100;
+        uint64 nextEpoch;
+
+        // Move forward to the exit epoch
+        nextEpoch = exitEpoch + slackForEscrowPeriod;
+        lightClient.setCurrentEpoch(nextEpoch);
+
+        uint256 userIndex = userIndexFromVk[vkHash];
+        address userAddress = users[userIndex];
+        vm.prank(userAddress);
+        stakeTable.withdrawFunds(vk);
+
+        // Remove element from array
+        requestExitKeys[index] = requestExitKeys[requestExitKeys.length - 1];
+        requestExitKeys.pop();
+        exitEpochForBlsVK[vkHash] = 0; // No exit in progress anymore
+    }
+}
+
+contract StakeTableInvariant_Tests is Test {
+    S public stakeTable;
+    ExampleToken public token;
+    LightClientTest public lightClientContract;
+    uint256 public constant INITIAL_BALANCE = 1_000_000_000_000;
+    address public exampleTokenCreator;
+    address[] public users;
+
+    StakeTableHandlerTest public handler;
+
+    function setUp() public {
+        exampleTokenCreator = makeAddr("tokenCreator");
+        vm.prank(exampleTokenCreator);
+        token = new ExampleToken(INITIAL_BALANCE);
+
+        LightClient.LightClientState memory genesis = LightClient.LightClientState({
+            viewNum: 0,
+            blockHeight: 0,
+            blockCommRoot: BN254.ScalarField.wrap(1),
+            feeLedgerComm: BN254.ScalarField.wrap(1),
+            stakeTableBlsKeyComm: BN254.ScalarField.wrap(1),
+            stakeTableSchnorrKeyComm: BN254.ScalarField.wrap(1),
+            stakeTableAmountComm: BN254.ScalarField.wrap(1),
+            threshold: 10
+        });
+        uint32 numBlocksPerEpoch = 10;
```

Comment:
i feel like if we were to set it to lower value, then we can test more "cross-epoch" behavior/scenarios.
can we use 4 instead?

Reply:
Yes, done in 548207f9432ce77e994bcaeda5f20b580e310c38.